services/spar/src/Spar/API.hs

```diff
@@ -353,21 +353,34 @@ validateNewIdP apiversion _idpMetadata teamId mReplaces = withDebugLog "validate
   SAML.logger SAML.Debug $ show (apiversion, _idpMetadata, teamId, mReplaces)
   SAML.logger SAML.Debug $ show (_idpId, oldIssuers, idp)
 
-  let -- FUTUREWORK: some database lookups here are unnecessary, this could be made more
-      -- efficient rather easily.
-      handleIdPClash :: Either SAML.IdPId IdP -> m ()
-      handleIdPClash (Right idp') = case apiversion of
-        WireIdPAPIV1 -> do
+  let handleIdPClash :: Either SAML.IdPId IdP -> m ()
+      handleIdPClash = case apiversion of
+        WireIdPAPIV1 -> const $ do
           throwSpar $ SparNewIdPAlreadyInUse "you can't create an IdP with api-version v1 if the issuer is already in use on the wire instance."
-        WireIdPAPIV2 -> do
-          when (fromMaybe defWireIdPAPIVersion (idp' ^. SAML.idpExtraInfo . wiApiVersion) == WireIdPAPIV1) $ do
-            throwSpar $ SparNewIdPAlreadyInUse "only allow all-new IdPs, no combination of old and new IdPs."
-          when ((idp' ^. SAML.idpExtraInfo . wiTeam) == teamId) $ do
-            throwSpar $ SparNewIdPAlreadyInUse "if the exisitng IdP is registered for a team, the new one can't have it."
-      handleIdPClash (Left id') = do
-        let err = throwSpar . SparIdPNotFound . cs . show $ id' -- database inconsistency
-        idp' <- wrapMonadClient (Data.getIdPConfig id') >>= maybe err pure
-        handleIdPClash (Right idp')
+        WireIdPAPIV2 -> \case
+          (Right idp') -> do
+            guardSameTeam idp'
+            guardReplaceeV2
+          (Left id') -> do
+            idp' <- do
+              let err = throwSpar $ SparIdPNotFound (cs $ show id') -- database inconsistency
+              wrapMonadClient (Data.getIdPConfig id') >>= maybe err pure
+            handleIdPClash (Right idp')
+
+      guardSameTeam :: IdP -> m ()
+      guardSameTeam idp' = do
+        when ((idp' ^. SAML.idpExtraInfo . wiTeam) == teamId) $ do
```

Comment:
i think this line is redundant, but that's not affecting behavior, so let's fix that in a different PR.

Reply:
we should also clarify: if this is a replacement, then the issuer of the replacee is different, so we wouldn't find the replacee here.

if we want to replace an idp with one that has the same issuer, we would use PUT instead of POST.

Reply:
28dba9c410809a9e20d660de84e488f67b799145, e2fdfdf4d1f9f6bab338dc360360a6d07fed8f5f